### PR TITLE
ORM3 compatibility

### PR DESCRIPTION
**Pull request by connorhu**

### CODE REVIEW

File: composer.json
Comment (derrabus) [inline composer.json:12]: Is it really necessary to drop PHP 7 support in order to add ORM 3 support?

Note that you did not adjust the CI: It's still configured for running tests on PHP 7.2, 7.3 and 7.4.

File: composer.json
Comment (derrabus) [inline composer.json:13]:
```suggestion
        "doctrine/orm": "^2.7 || ^3.0.0-beta2"
```

File: composer.json
Comment (derrabus) [inline composer.json:21]: Why are those these dependency changes needed for ORM 3 compatibility?

File: .github/workflows/tests.yml
Comment (derrabus) [inline .github/workflows/tests.yml:20]: This should be a separate PR.

File: src/Query/AbstractCryptParser.php
Comment (derrabus) [inline src/Query/AbstractCryptParser.php:9]: What do we need all those new abstract classes for?

**Review by derrabus — COMMENTED** [review]
Thank you for starting the work on this topic. I get the impression that you're trying to cover multiple topic at once. Try to keep your PR focused. For instance, dropping PHP 7 support or upgrading PHP CS Fixer should be separate PRs.

File: composer.json
Comment (connorhu) [reply composer.json:12]: Current code is not compatible with pre-8 versions due to `mixed` return type. If I set the return type correctly everywhere I think it can be compatible with earlier versions again.

File: src/Query/AbstractCryptParser.php
Comment (connorhu) [reply src/Query/AbstractCryptParser.php:9]: There was a lot of duplicated code, often copy paste and the property name was not correct (it was called date but did not store a date related node). Eliminate this duplication and introduce a more generic property name.

Comment (derrabus) [reply src/Query/AbstractCryptParser.php:9]: I understand that abstracting duplicated code and renaming properties might be desirable in general, but this way you're introducing a lot of unrelated changes which might render the code review difficult.

Comment (connorhu) [reply src/Query/AbstractCryptParser.php:9]: If you tell me that the separate PR will be merged and not get stuck in the queue of waiting PRs, I'll be happy to split this PR into 4 parts (which is a draft anyway, so I don't expect a review)

Comment (derrabus) [reply src/Query/AbstractCryptParser.php:9]: It's @beberlei's repository, so I can't make any promises on what will be merged and what not. But I would try to support you and provide reviews, so that it's easier for @beberlei to hit the merge button eventually. 🙂

Comment (derrabus) [reply src/Query/AbstractCryptParser.php:9]: > should there be a CI php8.3 test? yes / no

Yes, but PHP 8.3 is completely irrelevant for ORM 3 compatibility.

> should cs-fixer 3.x upgrade? yes / no

Probably, but again: Not relevant for ORM 3.

> should there be consistency in the method order. This can be guaranteed by cs-fixer, see `ordered_class_elements` fixer: properties, getSql method, parse method (I like parts that do the same thing to look consistent)? yes / no

I have no opinion on that, but it's super irrelevant for ORM 3.

> should code be simplified with abstract classes? yes / no

We could do that. No necessary for ORM 3 support though, imho.

> should run separate tests with ORM 2.x and ORM 3.x? yes / no

As long as both versions are supported, we need to test against both. However, we can be lazy here: ORM 2 will be tested on the PHP 7 jobs automatically because ORM 3 requires PHP 8.1.

> If you look at `$parser->match()` methods, because of the `LexerType` enum, it can only be done with local ifs. Using abstract classes also reduces the footprint of this change.

If you just need to share some glue code among classes, you can use a trait instead. That's what traits are made for.

Comment (connorhu) [reply src/Query/AbstractCryptParser.php:9]: > > If you look at `$parser->match()` methods, because of the `LexerType` enum, it can only be done with local ifs. Using abstract classes also reduces the footprint of this change.
> 
> If you just need to share some glue code among classes, you can use a trait instead. That's what traits are made for.

Interesting suggestion. I'll think about whether we could use traits to give a nicer solution to the `Lexer` contant vs `TokenType` enum/class problem. Where there is duplicate code, it's definitely a nicer solution, but the logic that only exists in one `FunctionNode` parser tends to make it harder to see and more work to throw ORM2 support in later.

**Comment by beberlei** [comment]
@connorhu ah, i merged a PR that does some improvements to Lexer compatibility already causing a few conflicts here. Can you integrate it or is that problematic?

**Comment by connorhu** [comment]
> @connorhu ah, i merged a PR that does some improvements to Lexer compatibility already causing a few conflicts here. Can you integrate it or is that problematic?

Sure, I can integrate it, because I've done those conversions too. There are still some open questions around code simplification, whether it makes sense or not, whether I should do it or not. I can make a table showing how much repetitive logic is in the code.

File: composer.json
Comment (derrabus) [inline composer.json:21]: Should be reverted.

File: composer.json
Comment (derrabus) [inline composer.json:38]: We don't need this anymore.

File: src/Query/Mysql/Acos.php
Comment (derrabus) [inline src/Query/Mysql/Acos.php:26]: @greg0ire looking at all this conditional code, would it make sense to backport `TokenType` to ORM 2.x?

File: src/Query/Mysql/AnyValue.php
Comment (derrabus) [inline src/Query/Mysql/AnyValue.php:20]: I would perform a forward feature detection instead.

```suggestion
        $orm3 = class_exsists(TokenType::class);
```

File: src/Types/ZendDateType.php
Comment (derrabus) [inline src/Types/ZendDateType.php:75]: Let's check for the new `ValueNotConvertible` instead.

File: tests/Query/DbTestCase.php
Comment (derrabus) [inline tests/Query/DbTestCase.php:28]: This check is bogus: The method exists in all ORM versions that we support.

File: tests/Query/DbTestCase.php
Comment (derrabus) [inline tests/Query/DbTestCase.php:41]: I don't understand what problem you're attempting to solve here.

File: tests/Query/DbTestCase.php
Comment (derrabus) [inline tests/Query/DbTestCase.php:74]: Please let's use feature detection instead.

File: tests/Query/Mysql/DateTest.php
Comment (derrabus) [inline tests/Query/Mysql/DateTest.php:73]: We don't support this ancient ORM version anymore.

**Review by derrabus — CHANGES_REQUESTED** [review]

File: tests/Query/Mysql/DateTest.php
Comment (derrabus) [reply tests/Query/Mysql/DateTest.php:73]: see #432

File: tests/Query/DbTestCase.php
Comment (connorhu) [reply tests/Query/DbTestCase.php]: I forgot to check since when the setMetadataCache and setQueryCache methods exist.

File: tests/Query/DbTestCase.php
Comment (connorhu) [reply tests/Query/DbTestCase.php:68]: If I understand correctly, the original purpose of the author was to be able to handle changes to the SQL parenthesis rendered from DQL. I followed this logic where there is a difference between ORM2 and ORM3.

Comment (derrabus) [reply tests/Query/DbTestCase.php:68]: Yes.

File: .github/workflows/orm2_tests.yml
Comment (derrabus) [inline .github/workflows/orm2_tests.yml:30]: Sorry, but it does not make sense to run all tests on PHP 7 against lowest deps only.

File: src/Types/CarbonDateTimeType.php
Comment (derrabus) [inline src/Types/CarbonDateTimeType.php:41]: Why are those new overrides necessary now?

File: tests/Query/DbTestCase.php
Comment (derrabus) [inline tests/Query/DbTestCase.php:24]: We should use attributes if they're available.

```suggestion
        if (PHP_VERSION_ID < 80000) {
```

File: tests/Query/Mysql/TrigTest.php
Comment (derrabus) [inline tests/Query/Mysql/TrigTest.php:77]: Please revert the changes to this file.

**Review by derrabus — CHANGES_REQUESTED** [review]

**Comment by derrabus** [comment]
Please revert the changes you've made to the CI. We don't need them.

File: src/Query/Mysql/Acos.php
Comment (greg0ire) [reply src/Query/Mysql/Acos.php:28]: I think it might, but `TokenType` is an enum, so there would be a type change from class to enum between 2 and 3, which is probably why I didn't introduce it in 2.x.

Comment (connorhu) [reply src/Query/Mysql/Acos.php:28]: @greg0ire Can `TokenType` in ORM2 be a class that becomes an enum in ORM3? The notation is similar, so nothing changes in the direction of using the library.

Comment (greg0ire) [reply src/Query/Mysql/Acos.php:28]: Maybe yes. We could add a private constructor or a constructor that always throws to prevent attempts at instantiating the class.

Comment (connorhu) [reply src/Query/Mysql/Acos.php]: @greg0ire @derrabus This solution really makes ORM2 and ORM3 support more user friendly.

The new class:
https://github.com/doctrine/orm/commit/bd0a198e12f0fbb3784f7ec70c1ca4b58ab90b36
And the usage of the class:
https://github.com/beberlei/DoctrineExtensions/blob/a5fba21f0c502591dc682370b546d8a679871eb8/src/Query/Mysql/Acos.php

Comment (derrabus) [reply src/Query/Mysql/Acos.php]: Please send a PR to the ORM with that new class.

Comment (connorhu) [reply src/Query/Mysql/Acos.php]: @derrabus Should I follow the instructions in the "How to choose the right branch" issue and do it in 2.19.x, or can I go down in version? It really doesn't affect anything if we include the class in earlier versions and it helps for example this library to support as many versions as possible.

**Comment by derrabus** [comment]
Whenever you have time:
* Please make PHPCS happy.
* Please revert everything in the .github directory.
* Please answer the remaining open question.

**Comment by connorhu** [comment]
To answer your question I have to figure out whats going on here with the carbon.

The types for Carbon were extend from `DateType`, and some versions did not run tests when combined. This could also mean that 1) some versions didn't work at all, just weren't tested, 2) it would be better to extend the `Type` class instead 3) the testing is too permissive/incorrect.

**Comment by connorhu** [comment]
Current implementation is not compatible with DBAL4.
When `convertToDatabaseValue` method of `TimeType`, `DateType` etc classes with DBAL<4 converts the value, it uses `DateTimeInterface` to check the type.
With DBAL4 it checks for `DateTime`.
DBA3:
https://github.com/doctrine/dbal/blob/de57e23f00c513263df93b9b10bbbb421bc9239f/src/Types/DateType.php#L49
DBAL4:
https://github.com/doctrine/dbal/blob/1a307d92f8531b0897176e95040051c7414d80a5/src/Types/DateType.php#L38

The problem here is that `CarbonImmutable*` types do not extend immutable types. If I fix this, then 4 tests still won't run because they try to save an non Immutable type in an Immutable field:
https://github.com/beberlei/DoctrineExtensions/blob/ffccbed8e5c7c8e906b383e3db13f061fef46a26/tests/Types/CarbonDateTest.php#L169

If we want these 4 tests to run, we need our own `convertToDatabaseValue` implementation.

**Comment by derrabus** [comment]
I see. Would #434 solve the problem?

**Comment by connorhu** [comment]
PR rebased, squasd into one commit. Now we have to wait for release ORM 2.19.0.

File: composer.json
Comment (derrabus) [inline composer.json:13]:
```suggestion
        "doctrine/orm": "^2.19 || ^3.0 @dev"
```

This should fix the CI on PHP 7 for the time being. We can revert it once ORM 2.19 has been released.

**Comment by connorhu** [comment]
Thank you for helping the process.

File: tests/Entities/ZendDate.php
Comment (tattali) [inline tests/Entities/ZendDate.php:16]:
```suggestion
    #[ORM\Column(type: 'integer')]
```

**Review by tattali — APPROVED** [review]

**Comment by FluffyDiscord** [comment]
composer.json needs to also be updated to allow orm 3+

**Comment by connorhu** [comment]
@Cool-Programmer When ORM 2.19 is released. It is unfortunate that the master is also the stable branch, but it is what it is.

**Comment by Cool-Programmer** [comment]
> @Cool-Programmer When ORM 2.19 is released. It is unfortunate that the master is also the stable branch, but it is what it is.

Hello, thank you very much for your reply.
So i'm getting such warnings in thousands:

[2024-02-29T09:21:59.392126+01:00] deprecation.INFO: User Deprecated: In ORM 3.0, the AttributeDriver will report fields for the classes where they are declared. This may uncover invalid mapping configurations. To opt into the new mode today, set the "reportFieldsWhereDeclared" constructor parameter to true. (AttributeDriver.php:78 called by App_KernelProdContainer.php:239, https://github.com/doctrine/orm/pull/10455, package doctrine/orm) {"exception":"[object] (ErrorException(code: 0): User Deprecated: In ORM 3.0, the AttributeDriver will report fields for the classes where they are declared. This may uncover invalid mapping configurations. To opt into the new mode today, set the \"reportFieldsWhereDeclared\" constructor parameter to true. (AttributeDriver.php:78 called by App_KernelProdContainer.php:239, https://github.com/doctrine/orm/pull/10455, package doctrine/orm) at /var/www/.................../vendor/doctrine/deprecations/lib/Doctrine/Deprecations/Deprecation.php:210)"} []

And when i update ORM to version 3, it is then incompatible with this DoctrineExtensions. What can we do?

**Comment by derrabus** [comment]
Please use https://github.com/doctrine/orm/discussions/ for ORM-related support questions.

**Comment by Cool-Programmer** [comment]
> Please use https://github.com/doctrine/orm/discussions/ for ORM-related support questions.

Hello,
May i ask how it is off topic? I think, I've clearly explained the situation. I'm getting that warning in thousands every single minutes, and i'm really not understanding how to proceed. Can you please give some info?

**Comment by connorhu** [comment]
@Cool-Programmer Your problem is caused by the fact that you have upgraded to ORM3. The error message refers to that, so you will get your answer in the doctrine discussion. This PR has not been merged yet, so the master branch of the repository is not compatible with ORM3.

**Comment by derrabus** [comment]
> Hello, May i ask how it is off topic?

Sure. This is a PR that makes a the DoctrineExtensions package compatible with the ORM 3. We are discussing the changes necessary to achieve this goal. That's the topic.

You're posting an excerpt from your deprecation log that you don't understand. That's a different topic.

I understand that you seek help and I told you where you can ask for it.

**Comment by abozhinov** [comment]
When to expect the new release?

**Comment by derrabus** [comment]
When it's ready.

**Comment by connorhu** [comment]
@derrabus I was thinking. It is a legitimate wish to use the code with orm3. Which is ready anyway. What if you make an orm3 branch, apply the contents of this PR as a patch, rewrite the requirement for orm3 only and put a note in the install description that until 2.19 is released, the ORM3 compatible code is there? If 2.19 is released, we will merge this PR, drop the orm3 the branch and leave the master only.

**Comment by derrabus** [comment]
I mean, we can merge the branch as is. It's ready anyway and Composer will take care that this package is not installed with ORM 2.18.

**Review by derrabus — APPROVED** [review]

**Comment by derrabus** [comment]
Ah well, we should've done a rebase before, to have PHPStan analyze your changes. I've pushed b6349f3305904efea37c9384b9a4ca5d6dc38ad6 in order to make the CI green.

**Comment by Chris53897** [comment]
Thanks. But we should not forget to remove the `@dev` after the stable release of doctrine/orm 2.19.
https://github.com/beberlei/DoctrineExtensions/blob/master/composer.json#L13

**Comment by derrabus** [comment]
Sure. Not that it has any impact downstream, but we will remove it.

**Comment by derrabus** [comment]
https://github.com/beberlei/DoctrineExtensions/releases/tag/v1.5.0

**Pull request merged**
